### PR TITLE
Added direct upload from video camera

### PR DESCRIPTION
**Pull request by timedin-de**
Regarding feature request #7358
- Adding Option to directly upload a video
- If camera permissions are not already granted, it will ask if for video or image after granting (its seems not not to be possible to get the camera-action after the permission request)
-  It seems that the videointent doesn't work on emulator? 
- Tested successfull with a real GooglePixel 6a

- [ ] Tests written, or not not needed

### CODE REVIEW

**Review by alperozturk96 — CHANGES_REQUESTED** [review]
I try to upload video directly on Android 14 emulator, after completing video shooting upload didn't work. The video file didn't upload.

After my second attempt, the app is crashed. Logcat didn't show any crash log for nextcloud package and general android

**Comment by timedin-de** [comment]
> Thanks for the contribution.
> 
> I try to upload video directly on Android 14 emulator, after completing video shooting upload didn't work. The video file didn't upload.
> 
> After my second attempt, the app is crashed. Logcat didn't show any crash log for nextcloud package and general android
> 
> Could you test via Android 14 emulator and share your findings with us?

Thanks for testing, thats what i meant with "It seems that the videointent doesn't work on emulator?", I tried countless times using the videointent on an emulator and it crashes or exits. I currently have no idea why this happens. On a hardware device it just works... (i was unsure if it was maybe just my underpowered laptop, but this ensures the bug)

I'll try some tests with only video intents on emulator to check for the issue.

**Comment by timedin-de** [comment]
Hey, I'm sorry, but I currently don't have the time (and motivation) to further check on this issue in the emulator. Maybe someone else finds the problem, or just discard the PR?

**Comment by alperozturk96** [comment]
> Hey, I'm sorry, but I currently don't have the time (and motivation) to further check on this issue in the emulator. Maybe someone else finds the problem, or just discard the PR?

I'll check it. Thank you for your contribution; it's wonderful to see support from the community 👍

**Comment by JonasMayerDev** [comment]
For me, the PR seems to work on real device (One Plus running Android 12).

**Comment by JonasMayerDev** [comment]
What do you think of not having 2 different uploads from camera in the Dialog Fragment but just one and then showing the user the dialog that's currently shown when accepting the permissions with the options video and image? I think the DialogFragment is already a little crowded, and it would make the experience a little more consistent since the dialog is currently only shown once regardless if "upload from camera" or "upload from video camera" is clicked...

@alperozturk96 @timedin-de What do you think?

**Comment by timedin-de** [comment]
> @timedin-de What do you think?

Yep I could do it this way if you agree on it.
It would mean one more click to open the camera, but i think would cause less confusion, (especially the current additional prompt after the permission granting). 
 
I chose the current approach because of the comment in the issue
_by @joshtrichards in https://github.com/nextcloud/android/issues/7358#issuecomment-1755486359_

**Comment by JonasMayerDev** [comment]
@timedin-de thank you! I guess the best persons to decide on that is the design team @nextcloud/designers 
To summarize for them:

Option 1 for adding a "capture video" implementation would look something like this:

image

Pro: One click video capture
Con: Too crowded dialog when clicking plus button, Inconsistent since dialog from option 2 would still show once when accepting video permissions.

Option 2 would look something like this:

Screenshot_2024-05-21-15-59-40-98_06ea656a9ba653ff96746f477f32d0f9

Pro: Consistent, less crowded
Con: 2 Clicks for video capture

**Comment by timedin-de** [comment]
> Honestly I dont unterstand what is the difference between upload from camera and upload from video camera?

On Android when a camera gets started it needs to be specified if it should record a video or take a picture (you cant decide/change it inside of the camera preview).

So "video camera" starts in video-recording-mode, "camera" in photo-mode

**Comment by JonasMayerDev** [comment]
@szaimen Yes good point. But the question is if we should use 2 separate buttons (video and image) when clicking on the plus in the app. Or if we combine it in one button and the ask the user in a dialog if wants video or image. Take a look at my message for the pro and cons in my opinion. https://github.com/nextcloud/android/pull/12995#issuecomment-2124541988

**Comment by szaimen** [comment]
I would probably indeed combine these two things into one option:
`Take photo or video and upload` and keep using the camera icon.

Then if clicking on that, indeed open a popup and ask this question but put the photo button to the right as it is more likely that you want to take a photo imho.

**Comment by timedin-de** [comment]
Ok I left it on "Upload from camera".

For implementation I moved the permissons-code from `FileOperationsHelper` to `OCFileListFragment`, I thought the dialog should be part of the Fragment not the FileHelper. Hit me up if I should change it.

File: app/src/main/java/com/owncloud/android/ui/activity/FileDisplayActivity.java [hunk 1]
```diff
@@ -443,7 +444,7 @@ public void onRequestPermissionsResult(int requestCode, @NonNull String[] permis
                 // If request is cancelled, result arrays are empty.
                 if (grantResults.length > 0 && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                     // permission was granted
-                    getFileOperationsHelper().uploadFromCamera(this, FileDisplayActivity.REQUEST_CODE__UPLOAD_FROM_CAMERA);
+                    getOCFileListFragmentFromFile().directCameraUpload();
```
Comment (timedin-de) [inline app/src/main/java/com/owncloud/android/ui/activity/FileDisplayActivity.java:447]: displaying the dialog in OCFileListFragment

File: app/src/main/java/com/owncloud/android/ui/activity/FileDisplayActivity.java [hunk 2]
```diff
@@ -844,17 +845,22 @@ public void onCheckAvailableSpaceFinish(boolean hasEnoughSpaceAvailable, String.
...
+                        File renamedFile;
+                        if(requestCode == REQUEST_CODE__UPLOAD_FROM_CAMERA) {
+                            renamedFile = new File(file.getParent() + PATH_SEPARATOR + FileOperationsHelper.getCapturedImageName());
+                        } else {
+                            renamedFile = new File(file.getParent() + PATH_SEPARATOR + FileOperationsHelper.getCapturedVideoName());
+                        }
 
                         if (!file.renameTo(renamedFile)) {
-                            DisplayUtils.showSnackMessage(getActivity(), "Fail to upload taken image!");
+                            DisplayUtils.showSnackMessage(getActivity(), R.string.error_uploading_direct_camera_upload);
```
Comment (timedin-de) [inline app/src/main/java/com/owncloud/android/ui/activity/FileDisplayActivity.java:856]: not regarding the video-upload itself, small improvement

File: app/src/main/java/com/owncloud/android/ui/fragment/OCFileListBottomSheetDialog.java [hunk 3]
```diff
@@ -75,12 +76,12 @@ protected void onCreate(Bundle savedInstanceState) {
...
-        viewThemeUtils.platform.colorImageView(binding.menuIconDirectCameraUpload);
-        viewThemeUtils.platform.colorImageView(binding.menuIconScanDocUpload);
-        viewThemeUtils.platform.colorImageView(binding.menuIconMkdir);
-        viewThemeUtils.platform.colorImageView(binding.menuIconAddFolderInfo);
+        viewThemeUtils.platform.colorImageView(binding.menuIconUploadFiles, ColorRole.PRIMARY);
+        viewThemeUtils.platform.colorImageView(binding.menuIconUploadFromApp, ColorRole.PRIMARY);
+        viewThemeUtils.platform.colorImageView(binding.menuIconDirectCameraUpload, ColorRole.PRIMARY);
+        viewThemeUtils.platform.colorImageView(binding.menuIconScanDocUpload, ColorRole.PRIMARY);
+        viewThemeUtils.platform.colorImageView(binding.menuIconMkdir, ColorRole.PRIMARY);
+        viewThemeUtils.platform.colorImageView(binding.menuIconAddFolderInfo, ColorRole.PRIMARY);
```
Comment (timedin-de) [inline app/src/main/java/com/owncloud/android/ui/fragment/OCFileListBottomSheetDialog.java:84]: Not regarding video upload. Fixing Deprecation

**Review by JonasMayerDev — APPROVED** [review]
Works for me on OnePlus Nord 2 and code looks good to me.

File: app/src/main/java/com/owncloud/android/ui/helpers/FileOperationsHelper.java [hunk 4]
```diff
@@ -1058,36 +1056,48 @@ public void checkCurrentCredentials(User user) {
         fileActivity.showLoadingDialog(fileActivity.getString(R.string.wait_checking_credentials));
     }
 
-    public void uploadFromCamera(Activity activity, int requestCode) {
-        Intent pictureIntent = new Intent(MediaStore.ACTION_IMAGE_CAPTURE);
+    public void uploadFromCamera(Activity activity, int requestCode, boolean isVideo) {
+        Intent cameraIntent;
+        if(isVideo) {
```
Comment (alperozturk96) [inline app/src/main/java/com/owncloud/android/ui/helpers/FileOperationsHelper.java:1061]:
```suggestion
        if (isVideo) {
```

**Comment by alperozturk96** [comment]
Thanks for the PR and contribution.

I just did some small refactoring and add FIXME note for uploading video from emulator. I can't push commit for this branch because it's not branched out from our master branch.

Could you please apply this git patch? Then we can merge it.

[refactor.patch](https://github.com/user-attachments/files/15882070/refactor.patch)

**Comment by timedin-de** [comment]
@alperozturk96 Thank you for your work and time.

Until now i haven't worked with patches, when i just apply it I get an error. Should I just undo my commit, apply the patch (with `git apply refactor.patch` or `git am < refactor.patch` and force-push or should i apply the patch on another branch (without my commit) and create a merge to the fork-master branch. 

And how do I have to handle signing off.

**Comment by JonasMayerDev** [comment]
@timedin-de not 100% sure but I would suggest number 1 ("remove your commits and add them via the patch back") since the patch seems to contain all changes you made compared to master. You can sign your commit with -S so eg. git commit -S -m "message" after you applied the patch

**Review by alperozturk96 — APPROVED** [review]

**Pull request merged**